### PR TITLE
Checks checksum consistency on backup chain

### PR DESCRIPTION
**Pull request by gustabowill**
Postgres incremental backups have the risk of the user switching the `data_checkums` configuration on the server between backups, which could potentially cause problems during recovery. An attempt of reducing these cases is implemented in this PR, where we will now store the state of `data_checksums` in the `backup.info` file of each backup, which will allow Barman to check the consistency of the chain during backup and recovery execution and emit warnings to the user so they can take action about it.

- 67b16d71e2f5a3bea23b18348031c46b56d5f406 Add the new `data_checksums` config to the `backup.info` file
- eb68c24f4da496686aa1d8788f17e76f012ea40c Fix tests broken due to the new config added in the previous commit
- b116e52d4e363ae01d58f9f959763dda876ccc1e Add a new method `is_checksum_consistent` in the `LocalBackupInfo` class to check for the chain consistency. This check is performed on backup and recovery executions.
- c29a4c7787ca74c1e6d85ca2311e8ac021b4e548 Add tests to the new method `is_checksum_consistent` created in the previous commit
- ef79368719bd301f79e29b5b965caf7846a80f79 Add the new checksums config to be displayed on the show-backup command. Related tests were also modified in this same commit, since they were very small changes.

### CODE REVIEW

File: barman/server.py [hunk 1]
```diff
@@ -1705,6 +1705,16 @@ def backup(self, wait=False, wait_timeout=None, backup_name=None, **kwargs):
...
+            # check if the backup chain (in case it is a Postgres incremental) is consistent
+            # with their checksums configurations
+            if not backup_info.is_checksum_consistent():
+                output.warning(
+                    "This is an incremental backup taken with `data_checksums = on` whereas "
+                    "some previous backups in the chain were taken with `data_checksums = off`. "
+                    "This can lead to potential recovery issues. Consider taking a new full backup "
+                    "to avoid having inconsistent backup chains."
+                )
+
```
Comment (gustabowill) [inline barman/server.py:1717]: I thought of putting the backup check on the `PostgresBackupExecutor`, since it is specifically for Postgres incremental backups, but I think the warning has more chances of not being ignored if displayed at the end, so I put it on the server backup method instead.

**Comment by barthisrael** [comment]
Just rebased on top of `master`.

File: barman/server.py (hunk 1 above)
Comment (barthisrael) [reply barman/server.py:1717]: I think that makes sense -- that way it's less prone to being missed/ignored.

File: barman/infofile.py [hunk 2]
```diff
@@ -945,6 +945,24 @@ def walk_to_root(self, return_self=True):
...
+    def is_checksum_consistent(self):
+        """
+        Check if all backups in the chain are consistent with their checksums
+        configurations.
+
+        The backup chain is considered inconsistent if the current backup was taken with
+        ``data_checksums`` enabled and any of its ascendants were taken with it disabled.
+        It is considered consistent otherwise.
+
+        :return bool: True if it is consistent, False otherwise.
```
Comment (barthisrael) [inline barman/infofile.py:957]:
```suggestion
        :return bool: ``True`` if it is consistent, ``False`` otherwise.
```

File: barman/infofile.py [hunk 3]
```diff
@@ -945,6 +945,24 @@ def walk_to_root(self, return_self=True):
...
 
+    def is_checksum_consistent(self):
+        """
+        Check if all backups in the chain are consistent with their checksums
+        configurations.
+
+        The backup chain is considered inconsistent if the current backup was taken with
+        ``data_checksums`` enabled and any of its ascendants were taken with it disabled.
+        It is considered consistent otherwise.
+
```
Comment (barthisrael) [inline barman/infofile.py:956]:
```suggestion

    .. note::
        While this method was created to check inconsistencies in chains of one
        or more (Postgres 17+ core) incremental backups, it can be safely used
        with any Postgres version and with any Barman backup method. That is
        true because it always returns ``True`` when called for a Postgres full
        backup or for a rsync backup.

```

I think we can add this note here to make this behavior explicit.
I guess it would also avoid questions in the future, if one finds weird the following statement at the end of the backup, which doesn't check the backup method nor the Postgres version:

```python
if not backup_info.is_checksum_consistent():
```

File: barman/recovery_executor.py [hunk 4]
```diff
@@ -1856,10 +1856,21 @@ def recover(self, backup_info, dest, remote_command=None, **kwargs):
...
             synthetic_backup_info, dest, remote_command=remote_command, **kwargs
         )
 
+        # If the checksum configuration is not consistent among all backups in the chain, we
+        # raise a warning at the end so the user can optionally take action about it
+        if not backup_info.is_checksum_consistent():
+            output.warning(
+                "You recovered from an incremental backup where checksums were enabled on "
+                "some, but not all backups in the chain. It is advised to disable, and "
+                "optionally reenable, checksums on the destination directory to avoid failures."
```
Comment (barthisrael) [inline barman/recovery_executor.py:1869]:
```suggestion
                "You recovered from an incremental backup where checksums were enabled on "
                "that backup, but not all backups in the chain. It is advised to disable, and "
                "optionally re-enable, checksums on the destination directory to avoid failures."
```

This sounds a bit "more accurate"?

File: tests/test_executor.py [hunk 5]
```diff
@@ -834,13 +834,15 @@ def test_exclusive_check(self, server_version, expected_message, capsys):
         ("server_version", "expected_value"),
         [(160000, None), (170000, "on")],
     )
-    def test__pg_get_metadata(self, server_version, expected_value):
+    def test__pg_get_metadata_summarize_wal(self, server_version, expected_value):
```
Comment (barthisrael) [inline tests/test_executor.py:837]:
```suggestion
    def test__pg_get_metadata(self, server_version, expected_value):
```

I'd say to revert this change.
I mean, we are testing more things in the test method.

File: tests/test_executor.py [hunk 6]
```diff
@@ -876,12 +878,17 @@ def test__pg_get_metadata(self, server_version, expected_value):
         mock_postgres.get_tablespaces.assert_called_once()
         mock_postgres.get_configuration_files.assert_called_once()
         if mock_postgres.server_version < 170000:
-            mock_postgres.get_setting.assert_called_once_with("data_directory")
+            calls = [call("data_directory"), call("data_checksums")]
+            mock_postgres.get_setting.assert_has_calls(calls, any_order=False)
```
Comment (barthisrael) [inline tests/test_executor.py:882]:
```suggestion
            mock_postgres.get_setting.assert_has_calls(calls, any_order=False)
            assert mock_postgres.get_setting.call_count == 2
```

To be even more precise 😄

File: barman/output.py [hunk 7]
```diff
@@ -853,6 +853,7 @@ def render_show_backup_base(backup_info, output_fun, header_row, nested_row):
         output_fun(nested_row.format("End Offset", backup_info["end_offset"]))
         output_fun(nested_row.format("Begin LSN", backup_info["begin_xlog"]))
         output_fun(nested_row.format("End LSN", backup_info["end_xlog"]))
+        output_fun(nested_row.format("Checksums", backup_info["data_checksums"]))
```
Comment (barthisrael) [inline barman/output.py:856]: We need to add this to the JSON output as well.

**Review by barthisrael — CHANGES_REQUESTED** [review]
Nice PR!

I still intend to perform some manual tests to check the outcomes as a user.

I put a few suggestions while reviewing it. Besides those code suggestions, I noticed we are missing a `References` in the last commit.

File: tests/test_executor.py (hunk 5 above)
Comment (barthisrael) [reply tests/test_executor.py:837]: We can remove the note from the second commit when reverting this change:

```
Additionally, one of the broken tests
was renamed from `test__pg_get_metadata` to
`test__pg_get_metadata_summarize_wal` as it was noticed that it was
specifically focused on testing a summarize wal config and not the whole
of the method.
```

File: barman/output.py (hunk 7 above)
Comment (gustabowill) [reply barman/output.py:856]: Oh True!

File: barman/recovery_executor.py (hunk 4 above)
Comment (gustabowill) [reply barman/recovery_executor.py:1869]: It does! The previous opens space for someone to think a backup with checksums disabled could also have its chain as inconsistent if any of the previous had it enabled, which is not true.

File: barman/infofile.py (hunk 3 above)
Comment (gustabowill) [reply barman/infofile.py:956]: Yeah, e.g. someone might inspecting the code specifically for rsync backups and get confused with this line at the end, so it's good to have an explanation like that so they know it is not meant for that case.

**Comment by barthisrael** [comment]
This also works as expected when recovering -- a `WARNING` is issued:

* List of backps:

```console
$ barman list-backup all
pg17 20240726T231540 'incremental_3' - Fri Jul 26 21:15:42 2024 - Size: 35.4 MiB - WAL Size: 0 B
pg17 20240726T231411 'incremental_2' - Fri Jul 26 21:14:13 2024 - Size: 35.4 MiB - WAL Size: 48.0 MiB
pg17 20240726T231359 'incremental_1' - Fri Jul 26 21:14:01 2024 - Size: 35.4 MiB - WAL Size: 32.0 MiB
pg17 20240726T231348 'full' - Fri Jul 26 21:13:51 2024 - Size: 251.3 MiB - WAL Size: 48.0 MiB
```

* State of backups in the chain:

```console
$ find ~/barman_home/ -name backup.info -print -exec bash -c "grep -i checksum {}; echo" \;
/home/vagrant/barman_home/pg17/base/20240726T231348/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T231359/backup.info
data_checksums=off

/home/vagrant/barman_home/pg17/base/20240726T231411/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T231540/backup.info
data_checksums=on
```

* `barman recover` attempt:

```console
$ barman recover pg17 incremental_3 ~/data_restore --recovery-staging-path /tmp
Start combining backup via pg_combinebackup for backup 20240726T231540 on /tmp
End combining backup via pg_combinebackup for backup 20240726T231540
Starting local restore for server pg17 using backup 20240726T231540
Destination directory: /home/vagrant/data_restore
Copying the base backup.
Copying required WAL segments.
Generating archive status files
Identify dangerous settings in destination directory.
WARNING: You recovered from an incremental backup where checksums were enabled on that backup, but not all backups in the chain. It is advised to disable, and optionally re-enable, checksums on the destination directory to avoid failures.

Recovery completed (start time: 2024-07-26 23:21:14.038575+02:00, elapsed time: less than one second)
Your PostgreSQL server has been successfully prepared for recovery!
```

* A similar messages can be found in the logs:

```console
$ tail -1 /var/log/barman/barman.log
2024-07-26 23:21:14,581 [34700] barman.recovery_executor WARNING: You recovered from an incremental backup where checksums were enabled on that backup, but not all backups in the chain. It is advised to disable, and optionally re-enable, checksums on the destination directory to avoid failures.
```

File: barman/output.py (hunk 7 above)
Comment (barthisrael) [reply barman/output.py:856]: Besides that, I'm wondering that we might want to move this to the "outer section", i.e. have something like this:

```
Backup 20240726T231540:
  Backup Name            : incremental_3
  Server Name            : pg17
  System Id              : 7382227563826631291
  Status                 : DONE
  PostgreSQL Version     : 170000
  PGDATA directory       : /var/lib/pgsql/17/data
  Checksums            : on
```

I suggest that only so we are consistent with the changes that were agreed in the call for Andre's PR about `show-backup`.

**Comment by barthisrael** [comment]
The previous tests were about cases where a `WARNING` was expected for backup and recovery commands, i.e. when the last backup has checksums enabled, but some backup in the chain does not.

I also tested the other cases, where there are divergences in the chain (mixes of `on` and `off`, but the latest backup is `off`), thus it's not expected that we get a `WARNING`. It went fine:

* For `barman backup`:

```console
# Get list of backups
$ barman list-backup all
pg17 20240726T231540 'incremental_3' - Fri Jul 26 21:15:42 2024 - Size: 35.4 MiB - WAL Size: 0 B
pg17 20240726T231411 'incremental_2' - Fri Jul 26 21:14:13 2024 - Size: 35.4 MiB - WAL Size: 48.0 MiB
pg17 20240726T231359 'incremental_1' - Fri Jul 26 21:14:01 2024 - Size: 35.4 MiB - WAL Size: 32.0 MiB
pg17 20240726T231348 'full' - Fri Jul 26 21:13:51 2024 - Size: 251.3 MiB - WAL Size: 48.0 MiB

# Check their checksum state
$ find ~/barman_home/ -name backup.info -print -exec bash -c "grep -i checksum {}; echo" \;
/home/vagrant/barman_home/pg17/base/20240726T231348/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T231359/backup.info
data_checksums=off

/home/vagrant/barman_home/pg17/base/20240726T231411/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T231540/backup.info
data_checksums=on

# Data checkums are disabled
$ psql -h localhost -U postgres -c "SHOW data_checksums"
 data_checksums
----------------
 off
(1 row)

# Take a new backup -- no WARNING about checkums, as expected
$ barman backup pg17 --name incremental_4 --incremental incremental_3
Starting backup using postgres method for server pg17 in /home/vagrant/barman_home/pg17/base/20240726T233256
Backup start at LSN: 6/7000110 (000000010000000600000007, 00000110)
Starting backup copy via pg_basebackup for 20240726T233256
Copy done (time: 1 second)
Finalising the backup.
Backup size: 3.4 MiB
Backup end at LSN: 6/9000000 (000000010000000600000009, 00000000)
Backup completed (start time: 2024-07-26 23:32:56.187060, elapsed time: 1 second)
Processing xlog segments from streaming for pg17
	000000010000000600000007
	000000010000000600000008
WARNING: IMPORTANT: this backup is classified as WAITING_FOR_WALS, meaning that Barman has not received yet all the required WAL files for the backup consistency.
This is a common behaviour in concurrent backup scenarios, and Barman automatically set the backup as DONE once all the required WAL files have been archived.
Hint: execute the backup command with '--wait'
```

* For `barman recover`:

```console
# List of backups
$ barman list-backup all
pg17 20240726T233256 'incremental_4' - Fri Jul 26 21:32:57 2024 - Size: 35.4 MiB - WAL Size: 0 B
pg17 20240726T231540 'incremental_3' - Fri Jul 26 21:15:42 2024 - Size: 35.4 MiB - WAL Size: 48.0 MiB
pg17 20240726T231411 'incremental_2' - Fri Jul 26 21:14:13 2024 - Size: 35.4 MiB - WAL Size: 48.0 MiB
pg17 20240726T231359 'incremental_1' - Fri Jul 26 21:14:01 2024 - Size: 35.4 MiB - WAL Size: 32.0 MiB
pg17 20240726T231348 'full' - Fri Jul 26 21:13:51 2024 - Size: 251.3 MiB - WAL Size: 48.0 MiB

# Their checksum state
$ find ~/barman_home/ -name backup.info -print -exec bash -c "grep -i checksum {}; echo" \;
/home/vagrant/barman_home/pg17/base/20240726T231348/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T231359/backup.info
data_checksums=off

/home/vagrant/barman_home/pg17/base/20240726T231411/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T231540/backup.info
data_checksums=on

/home/vagrant/barman_home/pg17/base/20240726T233256/backup.info
data_checksums=off

# Attempt to recover -- no WARNING, as expected
$ barman recover pg17 incremental_4 ~/data_restore --recovery-staging-path /tmp
Start combining backup via pg_combinebackup for backup 20240726T233256 on /tmp
End combining backup via pg_combinebackup for backup 20240726T233256
Starting local restore for server pg17 using backup 20240726T233256
Destination directory: /home/vagrant/data_restore
Copying the base backup.
Copying required WAL segments.
Generating archive status files
Identify dangerous settings in destination directory.

Recovery completed (start time: 2024-07-26 23:35:05.869706+02:00, elapsed time: less than one second)
Your PostgreSQL server has been successfully prepared for recovery!
```

File: barman/output.py (hunk 7 above)
Comment (gustabowill) [reply barman/output.py]: I applied this suggestion, but do you think it's worth considering adding a new section to this output? Considering that WAL Summarizer state will also come in another PR, maybe something like:

```
Backup 20240726T174335:
  Server Name            : pg17_server
  System Id              : 7376727553608684707
  Status                 : WAITING_FOR_WALS
  PostgreSQL Version     : 170000
  PGDATA directory       : /var/lib/pgsql/17/data

  Server Parameters:
      Checksums             :  on
      WAL Summarizer.   :  off  
    ....
    
```

I don't know. I don't think it's bad the way it is, just throwing some options on the table.

**Comment by gustabowill** [comment]
@barthisrael I just applied your suggestions. I also left a thought in [one of your replies](https://github.com/EnterpriseDB/barman/pull/963#discussion_r1693680429) about the output of `show-backup` command.

**Comment by barthisrael** [comment]
`barman show-backup` is also working, both in `console` and `json` writers:

```console
# Console writer
$ barman -f console show-backup pg17 incremental_4
Backup 20240726T233256:
  Backup Name            : incremental_4
  Server Name            : pg17
  System Id              : 7382227563826631291
  Status                 : DONE
  PostgreSQL Version     : 170000
  PGDATA directory       : /var/lib/pgsql/17/data
  Checksums              : off

  Tablespaces:
    custom_ts_1          : /var/lib/pgsql/17/tablespaces/custom_ts_1 (oid: 24678)
    custom_ts_2          : /var/lib/pgsql/17/tablespaces/custom_ts_2 (oid: 24686)

  Base backup information:
    Disk usage           : 3.4 MiB (35.4 MiB with WALs)
    Incremental size     : 3.4 MiB (-0.00%)
    Timeline             : 1
    Begin WAL            : 000000010000000600000008
    End WAL              : 000000010000000600000009
    WAL number           : 2
    Begin time           : 2024-07-26 21:32:56.183376+00:00
    End time             : 2024-07-26 21:32:57.553398+00:00
    Copy time            : 1 second
    Estimated throughput : 2.5 MiB/s
    Begin Offset         : 40
    End Offset           : 0
    Begin LSN            : 6/8000028
    End LSN              : 6/9000000

  WAL information:
    No of files          : 0
    Disk usage           : 0 B
    WAL rate             : 195.18/hour
    Last available       : 000000010000000600000009

  Catalog information:
    Retention Policy     : VALID
    Previous Backup      : 20240726T231540
    Next Backup          : - (this is the latest base backup)

# json writer
$ barman -f json show-backup pg17 incremental_4 | jq
{
  "pg17": {
    "backup_id": "20240726T233256",
    "backup_name": "incremental_4",
    "base_backup_information": {
      "analysis_time": "less than one second",
      "analysis_time_seconds": 0,
      "begin_lsn": "6/8000028",
      "begin_offset": 40,
      "begin_time": "2024-07-26 21:32:56.183376+00:00",
      "begin_time_timestamp": "1722029576",
      "begin_wal": "000000010000000600000008",
      "copy_time": "1 second",
      "copy_time_seconds": 1.350093,
      "disk_usage": "3.4 MiB",
      "disk_usage_bytes": 3528561,
      "disk_usage_with_wals": "35.4 MiB",
      "disk_usage_with_wals_bytes": 37082993,
      "end_lsn": "6/9000000",
      "end_offset": 0,
      "end_time": "2024-07-26 21:32:57.553398+00:00",
      "end_time_timestamp": "1722029577",
      "end_wal": "000000010000000600000009",
      "incremental_size": "3.4 MiB",
      "incremental_size_bytes": 3528561,
      "incremental_size_ratio": "-0.00%",
      "number_of_workers": 1,
      "throughput": "2.5 MiB/s",
      "throughput_bytes": 2613568.8430352574,
      "timeline": 1
    },
    "catalog_information": {
      "next_backup": "- (this is the latest base backup)",
      "previous_backup": "20240726T231540",
      "retention_policy": "VALID"
    },
    "data_checksums": "off",
    "pgdata_directory": "/var/lib/pgsql/17/data",
    "postgresql_version": 170000,
    "status": "DONE",
    "tablespaces": [
      {
        "location": "/var/lib/pgsql/17/tablespaces/custom_ts_1",
        "name": "custom_ts_1",
        "oid": 24678
      },
      {
        "location": "/var/lib/pgsql/17/tablespaces/custom_ts_2",
        "name": "custom_ts_2",
        "oid": 24686
      }
    ],
    "wal_information": {
      "compression_ratio": 0,
      "disk_usage": "0 B",
      "disk_usage_bytes": 0,
      "last_available": "000000010000000600000009",
      "no_of_files": 0,
      "timelines": [],
      "wal_rate": "195.18/hour",
      "wal_rate_per_second": 0.05421688016258301
    }
  }
}
```

File: barman/output.py (hunk 7 above)
Comment (barthisrael) [reply barman/output.py]: I think it's worth talking about that with the team.
Personally, I like your suggestion.

For now let's proceed as it is. If we decide to do that the way you suggested, we can do that at once through Andre's PR as that one will be merged later.

**Comment by barthisrael** [comment]
> @barthisrael I just applied your suggestions. I also left a thought in [one of your replies](https://github.com/EnterpriseDB/barman/pull/963#discussion_r1693680429) about the output of `show-backup` command.

I like your suggestion!
Let's bring that up to the team and get their thoughts.

**Review by barthisrael — APPROVED** [review]
Looks good!

File: barman/output.py (hunk 7 above)
Comment (andremagui) [reply barman/output.py]: I like having it separately like suggested!

Comment (barthisrael) [reply barman/output.py]: Martin mentioned to me he also thinks this is good.
